File: day:13/maxSubArray.py
class Solution:
    def isAllPositive(self, nums):
        for x in nums:
            if x < 0: return False
        return True
        
    def isAllNegative(self, nums):
        for x in nums:
            if x > 0: return False
        return True
    
    def maxSubArray(self, nums: List[int]) -> int:
        #base cases
        if(self.isAllPositive(nums)): return sum(nums);
        if(self.isAllNegative(nums)): return max(nums);
        
        sumSoFar = -inf;
        sumUpToIndex = 0;
    
        for num in nums:
            sumUpToIndex += num;
            sumUpToIndex = max(sumUpToIndex, 0);
            sumSoFar = max(sumSoFar, sumUpToIndex)
    
        return sumSoFar;

File: day:13/maxSubArray.java
class Solution {
    public boolean isAllPositive(int[] nums) {
        for(int x: nums) {
            if(x < 0) return false;
        }
        return true;
    }
    public boolean isAllNegative(int[] nums) {
        for(int x: nums) {
            if(x > 0) return false;
        }
        return true;
    }
    public int maxSubArray(int[] nums) {
        // base cases
        if(isAllPositive(nums)) return IntStream.of(nums).sum();
        if(isAllNegative(nums)) return IntStream.of(nums).max().getAsInt();

        int sumSoFar = Integer.MIN_VALUE;
        int sumUpToIndex = 0;
        for(int num: nums) {
            sumUpToIndex += num;
            sumUpToIndex = Math.max(sumUpToIndex, 0);
            sumSoFar = Math.max(sumSoFar, sumUpToIndex);
        }
        return sumSoFar;
        
    }
}
